app/src/org/commcare/utils/PrivilegesUtility.java

```diff
@@ -0,0 +1,185 @@
+        }
+    }
+
+    private Pair<String, String[]> processAndValidateV2Payload(JSONObject obj) throws JSONException, PrivilagePayloadException {
+        Pair<String[], String[]> payload = processV2Payload(obj);
+        if(validateV2PayloadSignature(payload.first[0], payload.first[1], payload.second)) {
+            return new Pair(payload.first[0], payload.second);
+        } else {
+            throw new InvalidPrivilegeSignatureException("Signatures don't match");
+        }
+    }
+
+    private boolean validateV2PayloadSignature(String username, String signature, String[] flags) throws JSONException, PrivilagePayloadException {
+        try {
+            byte[] signatureBytes = SigningUtil.getBytesFromString(signature);
+            String expectedUnsignedValue = getV2SignatureInput(flags, username);
+            return SigningUtil.verifyMessageAndBytes(authorityPublicKeyString, expectedUnsignedValue, signatureBytes) != null;
+        } catch (Exception e) {
+            throw new InvalidPrivilegeSignatureException(e);
+        }
+    }
+
+    private String getV2SignatureInput(String[] flags, String username) {
+        try {
+            JSONObject usernameObject = new JSONObject();
+            usernameObject.put("username", username);
+            JSONArray flagsArray = new JSONArray();
+            for(int i = 0 ; i < flags.length ; ++i) {
+                flagsArray.put(i, flags[i]);
+            }
+
+            JSONObject flagObject = new JSONObject();
+            flagObject.put("flags", flagsArray);
+
+            JSONArray array = new JSONArray();
+            array.put(usernameObject);
+            array.put(flagObject);
+            return array.toString();
+        } catch (JSONException e) {
+            return "";
+        }
+    }
+
+    private Pair<String[], String[]> processV2Payload(JSONObject obj) throws JSONException, PrivilagePayloadException {
+        String username = obj.getString("username");
+        JSONArray array = obj.getJSONArray("flags");
+        String[] permissions = new String[array.length()];
+        for(int i = 0 ; i < array.length(); ++i) {
+            permissions[i] = array.getString(i);
+        }
+
+        String signature = obj.getString("multiple_flags_signature");
+        return new Pair<>(new String[] {username, signature}, permissions);
+
+    }
+
+    private Pair<String, String[]> processAndValidateV1Payload(JSONObject obj) throws JSONException, PrivilagePayloadException {
+        String[] payload = processV1Payload(obj);
+        if(validateV1PayloadSignature(payload[0], payload[1], payload[2])) {
+            return new Pair(payload[1], new String[] {payload[0]});
+        } else {
+            throw new InvalidPrivilegeSignatureException("Signatures don't match");
+        }
+    }
+
+    private boolean validateV1PayloadSignature(String flag, String username, String signature) throws JSONException, PrivilagePayloadException {
```

Comment:
This method is identical to `validateV2PayloadSignature()` except for how `expectedUnsignedValue` is generated. Can you make that variable an argument to the method, and then the rest of the code need not be duplicated?